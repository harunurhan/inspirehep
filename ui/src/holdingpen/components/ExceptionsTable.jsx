import React, { Component } from 'react';
import { Table, Icon } from 'antd';
import PropTypes from 'prop-types';
import FilterDropdown from './FilterDropdown';
import './ExceptionsTable.scss';

class ExceptionsTable extends Component {
  static getDerivedStateFromProps(nextProps, prevState) {
    const { exceptions } = nextProps;
    return {
      ...prevState,
      allExceptions: exceptions,
      filteredExceptions: exceptions,
    };
  }

  constructor(props) {
    super(props);
    this.state = {
      selectedCollections: null,
      isErrorFilterDropdownVisible: false,
      isErrorFilterFocused: false,
      isRecidFilterDropdownVisible: false,
      isRecidFilterFocused: false,
    };

    this.onRecidFilterDropdownVisibleChange = this.onRecidFilterDropdownVisibleChange.bind(
      this
    );
    this.onErrorFilterDropdownVisibleChange = this.onErrorFilterDropdownVisibleChange.bind(
      this
    );
    this.onSelectedCollectionsChange = this.onSelectedCollectionsChange.bind(
      this
    );
    this.onErrorSearch = this.onErrorSearch.bind(this);
    this.onRecidSearch = this.onRecidSearch.bind(this);
    this.onColumnSearch = this.onColumnSearch.bind(this);
  }

  onErrorFilterDropdownVisibleChange(visible) {
    this.setState({
      isErrorFilterDropdownVisible: visible,
      isErrorFilterFocused: visible,
    });
  }

  onRecidFilterDropdownVisibleChange(visible) {
    this.setState({
      isRecidFilterDropdownVisible: visible,
      isRecidFilterFocused: visible,
    });
  }

  onErrorSearch(searchText) {
    const filteredExceptions = this.onColumnSearch(searchText, 'error');
    this.setState({
      isErrorFilterDropdownVisible: false,
      filteredExceptions,
    });
  }

  onRecidSearch(searchText) {
    const filteredExceptions = this.onColumnSearch(searchText, 'recid');
    this.setState({
      isRecidFilterDropdownVisible: false,
      filteredExceptions,
    });
  }

  onColumnSearch(searchText, columnToSearch) {
    const regExp = new RegExp(searchText, 'gi');
    return this.state.allExceptions.filter(exception => {
      const record = String(exception[columnToSearch]);
      return record.match(regExp);
    });
  }

  onSelectedCollectionsChange(_, filters) {
    this.setState({
      selectedCollections: filters,
    });
  }

  render() {
    const selectedCollections = this.state.selectedCollections || {};

    const columns = [
      {
        title: 'Collection',
        dataIndex: 'collection',
        filters: [
          { text: 'Hep', value: 'Hep' },
          { text: 'Hepnames', value: 'Hepnames' },
          { text: 'Job', value: 'Job' },
          { text: 'Jobhidden', value: 'Jobhidden' },
          { text: 'Conferences', value: 'Conferences' },
          { text: 'Journals', value: 'Journals' },
        ],
        filteredValue: selectedCollections.collection,
        onFilter: (value, record) => record.collection.includes(value),
        width: 120,
      },
      {
        title: 'Error',
        dataIndex: 'error',
        filterDropdown: (
          <FilterDropdown
            placeholder="Search error"
            onSearch={this.onErrorSearch}
            focused={this.state.isErrorFilterFocused}
          />
        ),
        filterIcon: <Icon type="search" />,
        filterDropdownVisible: this.state.isErrorFilterDropdownVisible,
        onFilterDropdownVisibleChange: visible => {
          this.onErrorFilterDropdownVisibleChange(visible);
        },
        width: '50%',
        render: text => text.split('\n', 1)[0],
      },
      {
        title: 'Record',
        dataIndex: 'recid',
        filterDropdown: (
          <FilterDropdown
            placeholder="Search recid"
            onSearch={this.onRecidSearch}
            focused={this.state.isRecidFilterFocused}
          />
        ),
        filterIcon: <Icon type="search" />,
        filterDropdownVisible: this.state.isRecidFilterDropdownVisible,
        onFilterDropdownVisibleChange: visible => {
          this.onRecidFilterDropdownVisibleChange(visible);
        },
        render: text => {
          const recordLink = `http://inspirehep.net/record/${text}/edit`;
          return <a href={recordLink}>{text}</a>;
        },
      },
    ];

    return (
      <Table
        className="__ExceptionsTable__"
        columns={columns}
        dataSource={this.state.filteredExceptions}
        rowKey="recid"
        pagination={{ pageSize: 25 }}
        onChange={this.onSelectedCollectionsChange}
        expandedRowRender={record => <pre>{record.error}</pre>}
        bordered
      />
    );
  }
}

ExceptionsTable.propTypes = {
  /* eslint-disable react/no-unused-prop-types */
  exceptions: PropTypes.arrayOf(
    PropTypes.shape({
      collection: PropTypes.string,
      error: PropTypes.string,
      recid: PropTypes.number,
    })
  ).isRequired,
  /* eslint-disable react/no-unused-prop-types */
};

export default ExceptionsTable;
