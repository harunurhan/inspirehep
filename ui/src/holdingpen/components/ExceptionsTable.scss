.__ExceptionsTable__ {
  width: 1400px;
}

@media (max-width: 1400px) {
  .__ExceptionsTable__ {
    width: 1200px;
  }
}

@media (max-width: 1200px) {
  .__ExceptionsTable__ {
    width: 800px;
  }
}
