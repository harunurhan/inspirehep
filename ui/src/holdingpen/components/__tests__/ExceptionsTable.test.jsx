import React from 'react';
import { shallow } from 'enzyme';
import ExceptionsTable from '../ExceptionsTable';

const exceptions = [
  {
    collection: 'Job',
    error:
      "Failed validating u'format' in schema[u'properties'][u'urls'][u'items'][u'properties'][u'value']:",
    recid: 1512550,
  },
  {
    collection: 'Hep',
    error:
      "Failed validating u'format' in schema[u'properties'][u'authors'][u'items'][u'properties'][u'emails'][u'items']:",
    recid: 1238165,
  },
  {
    collection: 'Conferences',
    error:
      "Failed validating u'format' in schema[u'properties'][u'urls'][u'items'][u'properties'][u'value']:",
    recid: 1356791,
  },
  {
    collection: 'Hep',
    error:
      "Failed validating u'pattern' in schema[u'properties'][u'persistent_identifiers'][u'items'][u'properties'][u'value']:",
    recid: 1635467,
  },
];

describe('ExceptionsTable', () => {
  it('renders with all props set', () => {
    const wrapper = shallow(<ExceptionsTable exceptions={exceptions} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should return correct filtered results', () => {
    const searchText = 'url';
    const wrapper = shallow(<ExceptionsTable exceptions={exceptions} />);
    const expectedFilteredResult = [
      {
        collection: 'Job',
        error:
          "Failed validating u'format' in schema[u'properties'][u'urls'][u'items'][u'properties'][u'value']:",
        recid: 1512550,
      },
      {
        collection: 'Conferences',
        error:
          "Failed validating u'format' in schema[u'properties'][u'urls'][u'items'][u'properties'][u'value']:",
        recid: 1356791,
      },
    ];
    const actualFilteredResult = wrapper
      .instance()
      .onColumnSearch(searchText, 'error');
    expect(actualFilteredResult).toEqual(expectedFilteredResult);
  });
});
