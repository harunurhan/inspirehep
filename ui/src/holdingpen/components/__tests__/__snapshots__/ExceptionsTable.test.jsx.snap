// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExceptionsTable renders with all props set 1`] = `
<Table
  bordered={true}
  className="__ExceptionsTable__"
  columns={
    Array [
      Object {
        "dataIndex": "collection",
        "filteredValue": undefined,
        "filters": Array [
          Object {
            "text": "Hep",
            "value": "Hep",
          },
          Object {
            "text": "Hepnames",
            "value": "Hepnames",
          },
          Object {
            "text": "Job",
            "value": "Job",
          },
          Object {
            "text": "Jobhidden",
            "value": "Jobhidden",
          },
          Object {
            "text": "Conferences",
            "value": "Conferences",
          },
          Object {
            "text": "Journals",
            "value": "Journals",
          },
        ],
        "onFilter": [Function],
        "title": "Collection",
        "width": 120,
      },
      Object {
        "dataIndex": "error",
        "filterDropdown": <FilterDropdown
          focused={false}
          onSearch={[Function]}
          placeholder="Search error"
    />,
        "filterDropdownVisible": false,
        "filterIcon": <Icon
          type="search"
    />,
        "onFilterDropdownVisibleChange": [Function],
        "render": [Function],
        "title": "Error",
        "width": "50%",
      },
      Object {
        "dataIndex": "recid",
        "filterDropdown": <FilterDropdown
          focused={false}
          onSearch={[Function]}
          placeholder="Search recid"
    />,
        "filterDropdownVisible": false,
        "filterIcon": <Icon
          type="search"
    />,
        "onFilterDropdownVisibleChange": [Function],
        "render": [Function],
        "title": "Record",
      },
    ]
  }
  dataSource={
    Array [
      Object {
        "collection": "Job",
        "error": "Failed validating u'format' in schema[u'properties'][u'urls'][u'items'][u'properties'][u'value']:",
        "recid": 1512550,
      },
      Object {
        "collection": "Hep",
        "error": "Failed validating u'format' in schema[u'properties'][u'authors'][u'items'][u'properties'][u'emails'][u'items']:",
        "recid": 1238165,
      },
      Object {
        "collection": "Conferences",
        "error": "Failed validating u'format' in schema[u'properties'][u'urls'][u'items'][u'properties'][u'value']:",
        "recid": 1356791,
      },
      Object {
        "collection": "Hep",
        "error": "Failed validating u'pattern' in schema[u'properties'][u'persistent_identifiers'][u'items'][u'properties'][u'value']:",
        "recid": 1635467,
      },
    ]
  }
  expandedRowRender={[Function]}
  indentSize={20}
  loading={false}
  locale={Object {}}
  onChange={[Function]}
  pagination={
    Object {
      "pageSize": 25,
    }
  }
  prefixCls="ant-table"
  rowKey="recid"
  rowSelection={null}
  showHeader={true}
  size="large"
  useFixedHeader={false}
/>
`;
